source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+
+      // Notify other processors how many particles we will send
+      MPI_Request *num_requests = new MPI_Request[2*num_neighbors];
+      for (unsigned int i=0; i<num_neighbors; ++i)
+        MPI_Irecv(&(num_recv_data[i]), 1, MPI_INT, neighbors[i], 0, this->get_mpi_communicator(), &(num_requests[2*i]));
+      for (unsigned int i=0; i<num_neighbors; ++i)
+        MPI_Isend(&(num_send_data[i]), 1, MPI_INT, neighbors[i], 0, this->get_mpi_communicator(), &(num_requests[2*i+1]));
+      MPI_Waitall(2*num_neighbors,num_requests,MPI_STATUSES_IGNORE);
+      delete[] num_requests;
+
+      // Determine how many particles and data we will receive
+      int total_recv_data = 0;
+      for (unsigned int neighbor_id=0; neighbor_id<num_neighbors; ++neighbor_id)
+        {
+          recv_offsets[neighbor_id] = total_recv_data;
+          total_recv_data += num_recv_data[neighbor_id];
+        }
+      const int num_recv_particles = total_recv_data / particle_size;
+
+      // Set up the space for the received particle data
+      std::vector<char> recv_data(total_recv_data);
+
+      // Exchange the particle data between domains
+      MPI_Request *requests = new MPI_Request[2*num_neighbors];
+      unsigned int send_ops = 0;
+      unsigned int recv_ops = 0;
+
+      for (unsigned int i=0; i<num_neighbors; ++i)
+        if (num_recv_data[i] > 0)
+          {
+            MPI_Irecv(&(recv_data[recv_offsets[i]]), num_recv_data[i], MPI_CHAR, neighbors[i], 1, this->get_mpi_communicator(),&(requests[send_ops]));
+            send_ops++;
+          }
+
+      for (unsigned int i=0; i<num_neighbors; ++i)
+        if (num_send_data[i] > 0)
+          {
+            MPI_Isend(&(send_data[send_offsets[i]]), num_send_data[i], MPI_CHAR, neighbors[i], 1, this->get_mpi_communicator(),&(requests[send_ops+recv_ops]));
+            recv_ops++;
+          }
+      MPI_Waitall(send_ops+recv_ops,requests,MPI_STATUSES_IGNORE);
+      delete[] requests;
+
+      // Put the received particles into the domain if they are in the triangulation
+      const void *recv_data_it = static_cast<const void *> (&recv_data.front());
+
+      for (int i=0; i<num_recv_particles; ++i)
+        {
+          const Particle<dim> recv_particle(recv_data_it,property_manager->get_particle_size());
+          integrator->read_data(recv_data_it, recv_particle.get_id());
+
+          typename parallel::distributed::Triangulation<dim>::active_cell_iterator cell;
+          try
+            {
+              cell = (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), recv_particle.get_location())).first;
+            }
+          catch (GridTools::ExcPointNotFound<dim> &)
+            {
+              // If we can find no cell for this particle it has left the domain due
+              // to an integration error or open boundary. Simply ignore the
+              // tracer in this case.
+              continue;
```

Comment:
Didn't the other processor determine that the particle is on one of our own cells? Does the case here happen due to round-off? Does it ever happen at all?